### PR TITLE
IDEM-1906: Only allow access to resource if idemeum entitlement exists (SSH Server & Win Desktops)

### PR DESCRIPTION
**Pull request by kpidemeum**

### CODE REVIEW

File: api/types/role.go [hunk 1]
```diff
@@ -290,6 +295,16 @@ func (r *RoleV5) SetOptions(options RoleOptions) {
 	r.Spec.Options = options
 }
 
+// GetLogins gets system logins for allow or deny condition.
+func (r *RoleV5) GetIdemeumEntitlements() []string {
+	return r.Spec.Allow.IdemeumEntitlements
+}
+
+// SetLogins sets system logins for allow or deny condition.
```
Comment (asyncid) [inline api/types/role.go:303]: Change the method comments its referring to logins

File: api/types/role.go [hunk 2]
```diff
@@ -290,6 +295,16 @@ func (r *RoleV5) SetOptions(options RoleOptions) {
 	r.Spec.Options = options
 }
 
+// GetLogins gets system logins for allow or deny condition.
```
Comment (asyncid) [inline api/types/role.go:298]: same here, change the comments

File: lib/services/role.go [hunk 3]
```diff
@@ -1829,45 +1829,56 @@ func (m *DatabaseNameMatcher) String() string {
...
-	logins := role.GetWindowsLogins(typ)
-	for _, login := range logins {
-		if l.login == login {
-			return true, nil
-		}
+	if l.entitlement == "" {
+		return matches(l.login, role.GetWindowsLogins(typ)), nil
 	}
-	return false, nil
+	return matches(l.login, role.GetWindowsLogins(typ)) && matches(l.entitlement, role.GetIdemeumEntitlements()), nil
```
Comment (asyncid) [inline lib/services/role.go:1881]: I don't think we can resolve the login names here, we are adding the user traits which is app_id+":"+names
You can only match against the idemeum entitlements here.

File: api/types/types.proto [hunk 4]
```diff
@@ -1789,6 +1789,9 @@ message CertExtension {
 // RoleConditions is a set of conditions that must all match to be allowed or
 // denied access.
 message RoleConditions {
+    // IdemeumEntitlements is a list of entitlements set by idemeum.
+    repeated string IdemeumEntitlements = 1000 [ (gogoproto.jsontag) = "idemeum_entitlements,omitempty" ];
```
Comment (asyncid) [inline api/types/types.proto:1793]: Why should add the IdemeumEntitlements at end of all the attributes if required you can leave may be margin of 20 fields, currently it has 22 fields only, margin of 1000 is too much.

Comment (kpidemeum) [reply api/types/types.proto:1793]: Resolving issue as we discussed and dropped issue.

File: lib/services/role.go (hunk 3 above)
Comment (kpidemeum) [reply lib/services/role.go:1881]: These login names come from traits, no resolution is happening.

File: lib/auth/clt.go [hunk 5]
```diff
@@ -1170,6 +1170,11 @@ func (c *Client) SearchSessionEvents(fromUTC, toUTC time.Time, limit int, order
 	return events, lastKey, nil
 }
 
+// GetEntitledNode returns a node by name only if user is entitled to it with login name.
+func (c *Client) GetEntitledNode(ctx context.Context, namespace string, name string, login string) (types.Server, error) {
+	return nil, nil
```
Comment (asyncid) [inline lib/auth/clt.go:1175]: If we are not implementing throw not implemented error

File: api/client/proto/authservice.proto [hunk 6]
```diff
@@ -1951,6 +1951,9 @@ service AuthService {
     // WatchEvents returns a new stream of cluster events
     rpc WatchEvents(Watch) returns (stream Event);
 
+    // GetEntitledNode returns a node by name only if user is entitled to it with login name.
+    rpc GetEntitledNode(types.EntitlementLookupRequest) returns (types.ServerV2);
```
Comment (asyncid) [inline api/client/proto/authservice.proto:1955]: Why we are adding new method when the existing checker can filter the node using the login.

**Review by asyncid — APPROVED** [review]

File: lib/auth/clt.go (hunk 5 above)
Comment (kpidemeum) [reply lib/auth/clt.go]: Implemented.

**Pull request merged**